File: addons/guide-cpp/src/modifiers/guide_modifier_8_way_direction.cpp
#include "guide_modifier_8_way_direction.h"

using namespace godot;

void GUIDEModifier8WayDirection::_bind_methods() {
    BIND_ENUM_CONSTANT(EAST); BIND_ENUM_CONSTANT(NORTH_EAST); BIND_ENUM_CONSTANT(NORTH); BIND_ENUM_CONSTANT(NORTH_WEST);
    BIND_ENUM_CONSTANT(WEST); BIND_ENUM_CONSTANT(SOUTH_WEST); BIND_ENUM_CONSTANT(SOUTH); BIND_ENUM_CONSTANT(SOUTH_EAST);

    ClassDB::bind_method(D_METHOD("get_direction"), &GUIDEModifier8WayDirection::get_direction);
    ClassDB::bind_method(D_METHOD("set_direction", "dir"), &GUIDEModifier8WayDirection::set_direction);
    ADD_PROPERTY(PropertyInfo(Variant::INT, "direction", PROPERTY_HINT_ENUM, "EAST,NORTH_EAST,NORTH,NORTH_WEST,WEST,SOUTH_WEST,SOUTH,SOUTH_EAST"), "set_direction", "get_direction");
}

Vector3 GUIDEModifier8WayDirection::_modify_input(Vector3 input, double delta, int value_type) const {
    if (!input.is_finite()) return Vector3(Math_INF, Math_INF, Math_INF);
    if (input.is_zero_approx()) return Vector3(0, 0, 0);

    double angle = Math::atan2(-input.y, input.x);
    int octant = (int)Math::round(8.0 * angle / (2.0 * Math_PI) + 8.0) % 8;
    return (octant == direction) ? Vector3(1, 0, 0) : Vector3(0, 0, 0);
}

bool GUIDEModifier8WayDirection::is_same_as(const Ref<GUIDEModifier> &other) const {
    Ref<GUIDEModifier8WayDirection> o = other;
    if (o.is_null()) return false;
    return o->get_direction() == direction;
}

File: addons/GuideCS/Modifiers/GUIDEModifier8WayDirection.cs
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEModifier8WayDirection : GUIDEModifier {
	public enum GuideDirection {
		East = 0,
		NorthEast = 1,
		North = 2,
		NorthWest = 3,
		West = 4,
		SouthWest = 5,
		South = 6,
		SouthEast = 7
	}

	[Export] public GuideDirection Direction { get; set; } = GuideDirection.East;

	public override bool IsSameAs(GUIDEModifier other) {
		return other is GUIDEModifier8WayDirection otherDir && Direction == otherDir.Direction;
	}

	public override Vector3 _ModifyInput(Vector3 input, float delta, GUIDEAction.GUIDEActionValueType valueType) {
		if (!input.IsFinite()) return Vector3.Inf;
		if (input.IsZeroApprox()) return Vector3.Zero;

		var angleRadians = Mathf.Atan2(-input.Y, input.X);
		var octant = Mathf.RoundToInt(8 * angleRadians / Mathf.Tau + 8) % 8;

		if (octant == (int)Direction) return Vector3.Right; // (1, 0, 0) indicating boolean true
		else return Vector3.Zero;
	}

	public override string _EditorName() => "8-way direction";

	public override string _EditorDescription() => "Converts a 2D input into a boolean that is true when the\ninput direction matches the selected direction. Note, that north is negative Y,\nbecause in Godot negative Y is up.";
}
